### PR TITLE
RustDoc: Fix bounds linking trait.Foo instead of traitalias.Foo

### LINKED ISSUE #84782: RustDoc: Links to trait aliases go to `trait.Foo.html` but the pages are `traitalias.Foo.html`
Found as part of my change in #84767.

CI build failure: https://github.com/rust-lang/rust/pull/84767/checks?check_run_id=2480989637#step:24:3504
```
std/iter/struct.Peekable.html:177: broken link - core/ops/trait.TryWhereOutputEquals.html
std/iter/struct.Once.html:13: broken link - core/ops/trait.TryWhereOutputEquals.html
std/iter/struct.Rev.html:14: broken link - core/ops/trait.TryWhereOutputEquals.html
...
```

Checking locally, the generated file is
```
/rust/build/x86_64-pc-windows-msvc/doc/core/ops/traitalias.TryWhereOutputEquals.html
```

Trait, as well as its uses, are in <https://github.com/rust-lang/rust/pull/84767/commits/ee4aa516ea5ce210e62fe3a16877610c28915f89#diff-b5736412f785f181ff3fcd768a2ea5f3bc1c6313af38a0094e5ef49c0e489c4fR223>

### PR DESCRIPTION
**Pull request by scottmcm**
Fixes #84782

The code was assuming `Trait` when adding bounds to the cache, so add a check on the DefId to see what its kind really is.

r? @jyn514 

Before:
image

After:
image

### CODE REVIEW

**Comment by jyn514** [comment]
My review todo list:
- audit all the wildcards in rustdoc
- open an issue for getting rid of separate HIR handling
- audit all uses of ItemKind::Trait to see if they should handle TraitAlias as well

**Review by GuillaumeGomez — APPROVED** [review]

**Review by jyn514 — COMMENTED** [review]
I haven't had time to do all the things on https://github.com/rust-lang/rust/pull/84811#issuecomment-830755737, but I think we can land this in the meantime. Just a couple nits.

File: src/librustdoc/formats/cache.rs [hunk 1]
```diff
@@ -350,26 +351,43 @@ impl<'a, 'tcx> DocFolder for CacheBuilder<'a, 'tcx> {
...
                 }
             }
             clean::PrimitiveItem(..) => {
                 self.cache.paths.insert(item.def_id, (self.cache.stack.clone(), item.type_()));
             }
 
-            _ => {}
+            clean::ExternCrateItem { .. }
+            | clean::ImportItem(..)
+            | clean::OpaqueTyItem(..)
```
[the remark this answers is not in the document]
Comment (scottmcm) [reply src/librustdoc/formats/cache.rs:377]: Oh, is this one TAIT?  I would have guessed https://doc.rust-lang.org/nightly/unstable-book/language-features/extern-types.html

But overall, I'd like to leave the match as-is for this PR, if that's ok.  Then someone else can audit them later and move them to different sections with comments about why they're ok to be skipped, or whatever.

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/formats/cache.rs:377]: Yup, that seems good to me.

File: src/librustdoc/clean/mod.rs [hunk 2]
```diff
@@ -157,7 +157,12 @@ impl Clean<GenericBound> for hir::GenericBound<'_> {
 impl Clean<Type> for (ty::TraitRef<'_>, &[TypeBinding]) {
     fn clean(&self, cx: &mut DocContext<'_>) -> Type {
         let (trait_ref, bounds) = *self;
-        inline::record_extern_fqn(cx, trait_ref.def_id, ItemType::Trait);
+        let kind = match cx.tcx.def_kind(trait_ref.def_id) {
+            DefKind::Trait => ItemType::Trait,
+            DefKind::TraitAlias => ItemType::TraitAlias,
+            other => bug!("`TraitRef` had unexpected kind {:?}", other),
+        };
```
[the remark this answers is not in the document]
Comment (scottmcm) [reply src/librustdoc/clean/mod.rs:164]: Ah, makes sense that there'd already be a mapping for this somewhere 👍

File: src/test/rustdoc/trait-alias-mention.rs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#![feature(trait_alias)]
+#![feature(ptr_metadata)]
+
+#![crate_name = "foo"]
+
+// @has foo/fn.this_never_panics.html '//a[@title="traitalias core::ptr::metadata::Thin"]' 'Thin'
```
[the remark this answers is not in the document]
Comment (scottmcm) [reply src/test/rustdoc/trait-alias-mention.rs]: Thanks for the pointers!

File: src/test/rustdoc/trait-alias-mention.rs [hunk 4]
```diff
@@ -1,9 +1,10 @@
...
 
 #![crate_name = "foo"]
 
-// @has foo/fn.this_never_panics.html '//a[@title="traitalias core::ptr::metadata::Thin"]' 'Thin'
-pub fn this_never_panics<T: std::ptr::Thin>() {
-    assert_eq!(std::mem::size_of::<&T>(), std::mem::size_of::<usize>())
+extern crate trait_alias_mention;
+
+// @has foo/fn.mention_alias_in_bounds.html '//a[@title="traitalias trait_alias_mention::SomeAlias"]' 'SomeAlias'
+pub fn mention_alias_in_bounds<T: trait_alias_mention::SomeAlias>() {
```
Comment (jyn514) [inline src/test/rustdoc/trait-alias-mention.rs:9]: Is this testing what you want it to? You want to make sure the link isn't broken, right?
```suggestion
// @has - '//a[@href="../trait_alias_mention/traitalias.SomeAlias.html"]'
pub fn mention_alias_in_bounds<T: trait_alias_mention::SomeAlias>() {
```

File: src/librustdoc/clean/mod.rs [hunk 5]
```diff
@@ -17,11 +17,11 @@ use rustc_hir::def::{CtorKind, DefKind, Res};
 use rustc_hir::def_id::{CrateNum, DefId, CRATE_DEF_INDEX, LOCAL_CRATE};
 use rustc_index::vec::{Idx, IndexVec};
 use rustc_infer::infer::region_constraints::{Constraint, RegionConstraintData};
-use rustc_middle::bug;
 use rustc_middle::middle::resolve_lifetime as rl;
 use rustc_middle::ty::fold::TypeFolder;
 use rustc_middle::ty::subst::{InternalSubsts, Subst};
 use rustc_middle::ty::{self, AdtKind, Lift, Ty, TyCtxt};
+use rustc_middle::{bug, span_bug};
```
Comment (scottmcm) [inline src/librustdoc/clean/mod.rs:24]: `tidy` insists that I move this down here :|

Comment (jyn514) [reply src/librustdoc/clean/mod.rs:24]: rustfmt is very annoying sometimes :( but this change is fine.

**Pull request merged**